sgkit/stats/aggregation.py

```diff
@@ -0,0 +1,53 @@
+import xarray as xr
+from xarray import DataArray, Dataset
+
+
+def allele_count(ds: Dataset) -> DataArray:
+    """Compute allele count from genotype calls.
+
+    Parameters
+    ----------
+    ds : Dataset
+        Genotype call dataset such as from
+        `sgkit.create_genotype_call_dataset`.
+
+    Returns
+    -------
+    variant/allele_count : DataArray
+        Allele counts with shape (variants, alleles) and values
+        corresponding to the number of non-missing occurrences
+        of each allele.
+
+    Examples
+    --------
+
+    >>> import sgkit as sg
+    >>> from sgkit.testing import simulate_genotype_call_dataset
+    >>> ds = simulate_genotype_call_dataset(n_variant=3, n_sample=2, seed=1)
+    # TODO: Use solution for https://github.com/pystatgen/sgkit/issues/37 here instead
+    >>> ds['call/genotype'].to_series().unstack().astype(str).apply('/'.join, axis=1).unstack()
+    samples 0   1
+    variants
+    0       1/0	1/0
+    1       1/0	1/1
+    2       0/1	1/0
+    3       0/0	0/0
+
+    >>> allele_count(ds)
+    <xarray.DataArray 'variant/allele_count' (variants: 4, alleles: 2)>
+    array([[2, 2],
+        [1, 3],
+        [2, 2],
+        [4, 0]])
+    Dimensions without coordinates: variants, alleles
+    """
+    # Count each allele index individually as a 1D vector and
+    # restack into new alleles dimension with same order
+    gt, mask = ds["call/genotype"], ds["call/genotype_mask"]
```

Comment:
Shouldn't these default names be parameters with defaults in the method signature? (plus the ones below)
And the expectation that `ds` should contain both genotype and genotype_mask should also be documented and also probably validated + descriptive `ValueError` if missing. This kind of `dataset` validation is likely to exists in most methods - so it would make sense to have a helper method for this so other functions can reuse the validation of inputs.

Reply:
I'm a -1/2 on that because:

1. Our ValueError wouldn't be more informative than the xarray KeyError there
2. The thrust of https://discourse.smadstatgen.org/t/xarray-conventions-for-variation-data/44/4 and https://github.com/pystatgen/sgkit/pull/10 was to hit one representation of a Dataset that all downstream methods could make assumptions about.  My early attempts to put some kind of type around the Dataset instances that would formalize those assumptions didn't make it in, but I still think some assumptions should be allowed
3. I think mentioning that the Dataset should come from `create_genotype_call_dataset` at least makes it clear to the user what assumptions are made by the function.  Since there's not a type to refer to, it would be nice to have a better way to express that than "A dataset from `sgkit.create_genotype_call_dataset`" but I'm not sure how else to put it. 

I think most libraries can get around this easily with subclasses/wrapper types for data containers, but since we're not doing that, I'm not sure what the right way to avoid all the boilerplate is (I think there will be a ton of it for those two arrays).  [docrep](https://github.com/Chilipp/docrep) looks like a possibility I guess (example: https://github.com/xgcm/xgcm/blob/master/xgcm/grid.py#L545).  I haven't used it but have you or anybody else tried something that reduces repetition in docstrings like that?

Reply:
Re: 1. Validation of user inputs is public API a common good practice, and throwing meaningful `ValueError` is also a standard over leaking errors/assumptions from the implementation. But if you are strongly against it, I won't block on this.

Re: 2. I see. Sorry to be late to the discussion - I do like the idea of encapsulation (especially over inheritance, which can be problematic over time) of Dataset/Array, but I don't want to restart that discussion, unless you believe there is a good reason to do that. I wouldn't be surprised if eventually we end up with encapsulation anyway, one pattern that I had a good experience with was sth like this:

```python

class SgKitDataset:
  def __init__(...):
    self.dataset: Dataset = ... # note this is intentionally a public member

  @classmethod
  def create_genotype_call_dataset(...):
    ...

  @classmethod
  def from_dataset(...):
    ...

  def useful_sgkit_dataset_fun(...):
    ...
```

This pattern allows to capture domain/lib specific API + expose the underlying data structure if needed for the users, overall is a good middle ground. We used this specifically for Scio: [here](https://github.com/spotify/scio/blob/master/scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala) to encapsulated Apache Beam's/Java `PCollection` with our Scala based `SCollection`, and it worked great in that case. I don't want to block your PR on this obviously, and if you like I could give it a try to sketch the API together in a PR given the functionalities currently in the code? It would mitigate issue from 2 and 3.

Reply:
> Re: 1. Validation of user inputs to public API is a common good practice, and throwing meaningful ValueError is also a standard over leaking errors/assumptions from the implementation. But if you are strongly against it, I won't block on this.

I think for just checking data variable presence, it's a grey area and I'd say no here but yes to that in an API like this if it happens: https://github.com/pystatgen/sgkit/issues/25#issuecomment-656093123.  But, I can definitely see the value in a set of checks specific to this Dataset that do more substantial things like check dimensions/dtypes and possibly assert things about values too (though the last one won't be great though without https://github.com/dask/dask/issues/97).  

> This pattern allows to capture domain/lib specific API + expose the underlying data structure if needed for the users (given the public field/accessor)

We wanted to avoid wrapper classes in the beginning and we talked about a bunch of other possibilities with subclasses and Xarray accessors.  My initial proposal was pretty similar to yours but tried to do some of it with [accessors](https://salem.readthedocs.io/en/latest/xarray_acc.html) instead since they behave much like a wrapper class.  Ultimately, @tomwhite was an advocate of using `Dataset`'s as-is in https://github.com/pystatgen/sgkit/pull/10 so that's the direction we've been heading.  I think it's a good one if we can iron out some of the boilerplate and I think there was also some general agreement that a validation function could be added to the top of every other function to replace what you and I are trying to do with types.  

A question I have for @tomwhite then is if a `check_genotype_call_dataset` function exists that we use at the beginning of every other analytical function, how much of the dataset do we validate?  On one hand it would be limiting to not be able to use `allele_count` because no `sample_id` variable exists (since it doesn't use it) but on the other it would be nice to enforce the same structure everywhere, like like a subclass or wrapper class would.  I'd defer to him on that.

Reply:
I agree that subclassing is more problematic than encapsulation, and we should consider the "middle ground" approach being suggested by @ravwojdyla. The advantage would be that checks are only needed on creation, rather than before every analytic function call.

So definitely worth sketching out such an API to see what we think of it. I'll add a few thoughts to #16 since that seems to be where the API discussion is taking place.

> if a check_genotype_call_dataset function exists that we use at the beginning of every other analytical function, how much of the dataset do we validate?

I was thinking that the validation would be lightweight - basically presence of variables, and checking dtypes.

Reply:
> I was thinking that the validation would be lightweight - basically presence of variables, and checking dtypes.

What I mean is what variables should be validated?  A function like `count_alleles` only needs some of them so checking all of them is limiting for using the function.  In other words, you don't need variables like`variant/contig` and `sample/id` so why should the `check_genotype_call_dataset` look for them?  

On the other hand, validating all the variables every time maintains more consistency with how `create_genotype_call_dataset` acts like a constructor for a wrapper or subclass type.

Reply:
I see. In that case I think you would only check for what you need.

Reply:
> 3\. Since there's not a type to refer to, it would be nice to have a better way to express that than "A dataset from `sgkit.create_genotype_call_dataset`" but I'm not sure how else to put it.

I'd suggest we write up a page of documentation which describes the xarray+sgkit dataset convention. That would describe what arrays (variables) and dimensions are expected, expected constraints on data types and dimensions, any naming conventions, etc.

Reply:
@alimanfoo sounds great! want to file an issue and add the `documentation` label?

Reply:
> @alimanfoo sounds great! want to file an issue and add the `documentation` label?

Done: #42